engines/vcruise/runtime.cpp

```diff
@@ -0,0 +1,1342 @@
+		changeAnimation(_idleAnimations[_direction]);
+		_animPlayWhileIdle = true;
+	}
+
+	_idleIsOnInteraction = false;
+
+	detectPanoramaDirections();
+
+	changeToCursor(_cursors[kCursorArrow]);
+	dischargeIdleMouseMove();
+}
+
+void Runtime::changeToCursor(const Common::SharedPtr<Graphics::WinCursorGroup> &cursor) {
+	if (!cursor)
+		CursorMan.showMouse(false);
+	else {
+		CursorMan.replaceCursor(cursor->cursors[0].cursor);
+		CursorMan.showMouse(true);
+	}
+}
+
+void Runtime::dischargeIdleMouseMove() {
+	const MapScreenDirectionDef *sdDef = _map.getScreenDirection(_screenNumber, _direction);
+
+	Common::Point relMouse(_mousePos.x - _gameSection.rect.left, _mousePos.y - _gameSection.rect.top);
+
+	bool isOnInteraction = false;
+	uint interactionID = 0;
+	if (sdDef) {
+		for (const InteractionDef &idef : sdDef->interactions) {
+			if (idef.rect.contains(relMouse)) {
+				isOnInteraction = true;
+				interactionID = idef.interactionID;
+				break;
+			}
+		}
+	}
+
+	if (isOnInteraction && (_idleIsOnInteraction == false || interactionID != _idleInteractionID)) {
+		_idleIsOnInteraction = true;
+		_idleInteractionID = interactionID;
+
+		// New interaction, is there a script?
+		Common::SharedPtr<Script> script = findScriptForInteraction(interactionID);
+
+		if (script)
+			activateScript(script, ScriptEnvironmentVars());
+	} else if (!isOnInteraction && _idleIsOnInteraction) {
+		_idleIsOnInteraction = false;
+		changeToCursor(_cursors[kCursorArrow]);
+	}
+}
+
+void Runtime::loadMap(Common::SeekableReadStream *stream) {
+	byte screenDefOffsets[MapDef::kNumScreens * kNumDirections * 4];
+
+	if (!stream->seek(16))
+		error("Error skipping map file header");
+
+	if (stream->read(screenDefOffsets, sizeof(screenDefOffsets)) != sizeof(screenDefOffsets))
+		error("Error reading map offset table");
+
+	for (uint screen = 0; screen < MapDef::kNumScreens; screen++) {
+		for (uint direction = 0; direction < kNumDirections; direction++) {
+			uint32 offset = READ_LE_UINT32(screenDefOffsets + (kNumDirections * screen + direction) * 4);
+			if (!offset)
+				continue;
+
+			if (!stream->seek(offset))
+				error("Error seeking to screen data");
+
+			byte screenDefHeader[16];
+			if (stream->read(screenDefHeader, 16) != 16)
+				error("Error reading screen def header");
+
+			uint16 numInteractions = READ_LE_UINT16(screenDefHeader + 0);
+
+			if (numInteractions > 0) {
+				Common::SharedPtr<MapScreenDirectionDef> screenDirectionDef(new MapScreenDirectionDef());
+				screenDirectionDef->interactions.resize(numInteractions);
+
+				for (uint i = 0; i < numInteractions; i++) {
+					InteractionDef &idef = screenDirectionDef->interactions[i];
+
+					byte interactionData[12];
+					if (stream->read(interactionData, 12) != 12)
+						error("Error reading interaction data");
+
+					idef.rect = Common::Rect(READ_LE_INT16(interactionData + 0), READ_LE_INT16(interactionData + 2), READ_LE_INT16(interactionData + 4), READ_LE_INT16(interactionData + 6));
```

Comment:
Why not use here a more intuitive `stream->readUint16LE()` for every variable?

Reply:
C++ order of parameter evaluation is undefined (which I have been burned by before - VS optimizer can reorder execution of parameter expressions even if the expressions have side effects) so it would need to be separate read lines, plus I think this way makes it easier to detect a read failure.